runners/local-artifact-service-java/src/test/java/org/apache/beam/artifact/local/LocalFileSystemArtifactStagerServiceTest.java

```diff
@@ -0,0 +1,222 @@
+
+    String name = "my-artifact";
+    requestObserver.onNext(PutArtifactRequest.newBuilder().setName(name).build());
+    requestObserver.onNext(
+        PutArtifactRequest.newBuilder()
+            .setData(ArtifactChunk.newBuilder().setData(ByteString.copyFrom(data)).build())
+            .build());
+    requestObserver.onCompleted();
+
+    String[] recordedFiles = stagingLocation.list();
+    Assert.assertThat(recordedFiles, Matchers.arrayWithSize(1));
+    File staged = stagingLocation.toPath().resolve(recordedFiles[0]).toFile();
+    ByteBuffer buf = ByteBuffer.allocate(data.length);
+    new FileInputStream(staged).getChannel().read(buf);
+    Assert.assertArrayEquals(data, buf.array());
+  }
+
+  @Test
+  public void multiPartPutArtifactSucceeds() throws Exception {
+    byte[] partOne = "foo-".getBytes();
+    byte[] partTwo = "bar-".getBytes();
+    byte[] partThree = "baz".getBytes();
+    RecordingStreamObserver<PutArtifactResponse> responseObserver = new RecordingStreamObserver<>();
+    StreamObserver<PutArtifactRequest> requestObserver = stager.putArtifact(responseObserver);
+
+    String name = "my-artifact";
+    requestObserver.onNext(PutArtifactRequest.newBuilder().setName(name).build());
+    requestObserver.onNext(
+        PutArtifactRequest.newBuilder()
+            .setData(ArtifactChunk.newBuilder().setData(ByteString.copyFrom(partOne)).build())
+            .build());
+    requestObserver.onNext(
+        PutArtifactRequest.newBuilder()
+            .setData(ArtifactChunk.newBuilder().setData(ByteString.copyFrom(partTwo)).build())
+            .build());
+    requestObserver.onNext(
+        PutArtifactRequest.newBuilder()
+            .setData(ArtifactChunk.newBuilder().setData(ByteString.copyFrom(partThree)).build())
+            .build());
+    requestObserver.onCompleted();
+
+    String[] recordedFiles = stagingLocation.list();
+    Assert.assertThat(recordedFiles, Matchers.arrayWithSize(1));
+    File staged = stagingLocation.toPath().resolve(recordedFiles[0]).toFile();
+    ByteBuffer buf = ByteBuffer.allocate("foo-bar-baz".length());
+    new FileInputStream(staged).getChannel().read(buf);
+    Assert.assertArrayEquals("foo-bar-baz".getBytes(), buf.array());
+  }
+
+  @Test
+  public void putArtifactBeforeNameFails() {
+    byte[] data = "foo-".getBytes();
+    RecordingStreamObserver<PutArtifactResponse> responseObserver = new RecordingStreamObserver<>();
+    StreamObserver<PutArtifactRequest> requestObserver = stager.putArtifact(responseObserver);
+
+    try {
+      requestObserver.onNext(
+          PutArtifactRequest.newBuilder()
+              .setData(ArtifactChunk.newBuilder().setData(ByteString.copyFrom(data)).build())
+              .build());
+      Assert.fail("Should throw out of the call to 'onNext'");
+    } catch (Exception ignored) {
+    }
+    Assert.assertThat(responseObserver.error, Matchers.not(Matchers.nullValue()));
+  }
+
+  @Test
+  public void putArtifactWithNoContentFails() {
```

Comment:
<!--new_thread; commit:fd3270d35ab3f8121ad44c2521dd0ade4223d84c; resolved:0-->
Does this need to fail?

Reply:
This is "Did not populate the request". We can accept a request with content but with an empty chunk, but we require that users send either the name or a chunk.